### PR TITLE
Incompatible mods detection

### PR DESCRIPTION
**Pull request by krzychu124**
With regards to #91 

### Incompatible mods detector.
- detects incompatible mods after game start
- displays popup with list of incompatible mods
- user can unsubscribe incompatible mod directly from this list without additional actions like opening Content Manager/Steam Workshop, game restarting etc.

__Further development__
- check for offline mods (User AppData folder)

### CODE REVIEW

**Comment by aubergine10** [comment]
testing...

**Comment by aubergine10** [comment]
In terms of functionality, it seems to work but I want to test a few more things...

File: TLM/TLM/UI/IncompatibleModsPanel.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            }
+        }
+
+        public Dictionary<ulong, string> IncompatibleMods { get; set; }
+
+        public void Initialize() {
+            Log._Debug("IncompatibleModsPanel initialize");
+            isVisible = true;
+
+            backgroundSprite = "MenuPanel3";
```
Comment (krzychu124) [inline TLM/TLM/UI/IncompatibleModsPanel.cs]: Change sprite at this line for different background

Comment (aubergine10) [reply TLM/TLM/UI/IncompatibleModsPanel.cs]: iterating various panel bgs, brb...

**Comment by aubergine10** [comment]
Noticed a bug while testing panel bgs....

If I "Esc" out of the warning dialog, I can no longer Esc out of the game (ie. quit to desktop from main menu by pressing Esc).

**Comment by krzychu124** [comment]
Sometimes warning dialog losing focus and you can't "Esc" it.
Was that warning closed or not?

File: TLM/TLM/UI/IncompatibleModsPanel.cs (hunk 1 above)
Comment (aubergine10) [reply TLM/TLM/UI/IncompatibleModsPanel.cs]: * "InfoDisplay" -- solid black
* "MenuContainer" -- midnight blue gradient
* "MenuPanel2" -- solid navy blue
* "UnlockingPanel2" -- sort of a mix between previous two

IMO **```UnlockingPanel2```** is the best one - it's distinct, so users won't mistake it for occasional "New DLC Features" panel, and it's fully opaque  so very easy to read the contents.

**Comment by aubergine10** [comment]
See edits to my earlier comment above.

Steps to reproduce:

1. Sub and enable TMPE (or, rather, have the build version enabled)
2. Quit cities.exe
3. Sub some incompat mods
4. Launch cities.exe - warning shows after pdx logo screen
5. Click "x" close button on warning panel

What happens:

* Screen still blurred
* I can no longer Esc to "Exit Game"
* I can, however, still use mouse to click main menu items

I'll do some testing of what happens when the warning panel is triggered from mods screen.

**Comment by aubergine10** [comment]
Everything works fine from mods screen triggering of panel (ie. when there are incompat mods subbed and TMPE is enabled inside mods screen, everything works good).

So it's only when launching game, with TMPE already enabled, and some incompat mods subbed, that the problem occurs.

Workaround: Opening mods screen (using mouse to click Content Manager link on main menu) restores normal operation (blurring goes, and Esc starts working).

**Comment by aubergine10** [comment]
Oh, I've somehow made the Unsubscribe buttons stop working when warning panel appears via mods screen (after enabling TMPE in content manager)... They disable, but don't disappear.

**Comment by krzychu124** [comment]
Damn... I have no idea how to fix that problem with blur...
To unblur you can even click exit then esc 😄 (The problem is that every vanilla dialog is made from dedicated dynamicPanel, but mine is just UIPanel dropped on top of popup stack)
Would be great if I could subscribe to result of this paradox login dialog so then I could run mods checker.

> Oh, I've somehow made the Unsubscribe buttons stop working when warning panel appears via mods screen (after enabling TMPE in content manager)... They disable, but don't disappear.

Tell me how? Could you reproduce it?

File: TLM/TLM/UI/IncompatibleModsPanel.cs (hunk 1 above)
Comment (krzychu124) [reply TLM/TLM/UI/IncompatibleModsPanel.cs]: Yup, ```UnlockingPanel2``` looks great

**Comment by aubergine10** [comment]
Hrm, I'm no longer getting first problem :/ (blurring)

**Comment by aubergine10** [comment]
See edited comment above for how to recreate second problem (buttons not disappear)

I'll do some digging in to first problem see if I can find a fix

**Comment by krzychu124** [comment]
> Hrm, I'm no longer getting first problem :/ (blurring)

Haha, me too because now I'm getting warning in mod tools ```m_SteamUGCRequestMap error: 15```

**Comment by aubergine10** [comment]
Ok, I know how to fix blurring issue :)

Assuming you can detect what screen the TMPE incompat mods warning is generated from (ie. main menu vs. content manager).

The blurring is caused by ```ModalEffect``` UITextureSprite

So if you get a handle to that, you can set it's ```.isVisible``` to false when the TMPE warning dlg closes.

When TMPE warning dlg is shown on game launch, you could potentially even set the .isVisible = true to get a nice blurred bg too :)

**Comment by aubergine10** [comment]
The blurring issue seems to happen when. on main menu screen, you press Esc while the TMPE warning is visible.

**Comment by krzychu124** [comment]
Isn't that texture attached to parent(caller) while dialog is shown(by dialog itself)?

**Comment by aubergine10** [comment]
I assume ```ModalEffect``` is present from game launch, and just has it's visibility toggled by COs UI stuff.

I tested using mod tools toggling ```.isVisible``` and it seemed to work fine.

Something like this should get handle to it:

```
 UIComponent component = GameObject.Find("ModalEffect").GetComponent<UIComponent>();

  if (component != null)
  {
      component.isVisible = false;
  }
```

**Comment by krzychu124** [comment]
Thanks, I will try to experiment with that.

**Comment by aubergine10** [comment]
An possibly better way would be to find some way to soak up the Esc keypress while the warning dialog is visible and stop it propagating to the  main menu.

**Comment by krzychu124** [comment]
Haha, I know what is going on with not working unsub button...
Instance is not cleared properly and it's stacking the same content at the same place. You can easly see it using DebugRender from ModTools under Ctrl + R ang Ctrl + G to cycle through layers from top to bottom

**Comment by aubergine10** [comment]
Did I mention that the UI library in C:SL is a nightmare? lol

**Comment by krzychu124** [comment]
I think we should add some wrappers to speed up building panels and other components

**Comment by aubergine10** [comment]
Agreed. That's kinda where I'm heading with my test mod for new UI. I want really clean, simple wrappers that just do what they're told without loads of titting around in crufty CO UI lib. If I was more experienced C# coder, I'd even consider writing a whole new UI library without about 90% less code than what CO has.

**Comment by krzychu124** [comment]
I didn't know that color property affects background sprite

zrzut ekranu 114

**Comment by aubergine10** [comment]
Yeah, I think it just colorises it.

Testing new updates...

**Comment by krzychu124** [comment]
I've managed to fix that blur issues, now testing blur while opened from content manager

**Comment by aubergine10** [comment]
LGTM

Blurring on main menu seems solved (Esc closes warning dlg). Double buttons on content manager solved (buttons now disappear when unsubbed). I'm not noticing any blur problems on content manager either (or is it that there is no blur when there should be?).

**Comment by krzychu124** [comment]
😕 I'm still seeing invisible panels are not destroyed properly

File: TLM/TLM/UI/IncompatibleModsPanel.cs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        }
+
+        public Dictionary<ulong, string> IncompatibleMods { get; set; }
+
+        public void Initialize() {
+            Log._Debug("IncompatibleModsPanel initialize");
+            if (mainPanel != null) {
+                Destroy(mainPanel);
+            }
```
Comment (aubergine10) [inline TLM/TLM/UI/IncompatibleModsPanel.cs]:
```
            if (mainPanel != null) {
                Destroy(mainPanel);
                _instance = null; // <-- does it need this?
            }
```

**Comment by aubergine10** [comment]
Which invisible panels are not destroyed?

**Comment by krzychu124** [comment]
Something is broken because I thought if we destroy parent then it will destroy their  children but it seems not...
Currently I can see that ```ScrollablePanel``` still has children but null parent after destroying ```mainPanel```

I will refactor it 😉

**Comment by aubergine10** [comment]
```OnDestroy()``` in ``UIComponent``` class should nuke the children.

**Comment by krzychu124** [comment]
Yup, I'm testing it now 😉

**Comment by krzychu124** [comment]
... I can't replace ```Destroy(mainPanel)``` with ```mainPanel.OnDestroy()``` because it gives me Infinite loop 😄

**Comment by krzychu124** [comment]
I got this, everything fixed, I think 😄

**Comment by krzychu124** [comment]
haha, I didn't commit that yet 😄

**Comment by aubergine10** [comment]
Urg, that wasn't the 'approve' I was expecting. I assume it'll go back to pending review if more changes are sent?

**Pull request merged**